### PR TITLE
Use smart pointer to manage pcfstat object

### PR DESCRIPTION
**Pull request by eryugey**
Previously pcfstat points to a raw pointer, and it may be leaked if
function returned before deleting it.

So use smart pointer to automatically release the object.

Note that currently s3fs only uses c++03, so we use auto_ptr here, not
unique_ptr, which requires c++11.

Fixes: https://github.com/s3fs-fuse/s3fs-fuse/commit/6ca5a24a7f29718967de1b8cc78dfe68c839729a ("Fix two inconsistency issues between stat cache and cache file (https://github.com/s3fs-fuse/s3fs-fuse/pull/2152)")
Signed-off-by: Eryu Guan <eguan@linux.alibaba.com>

### CODE REVIEW

**Comment by gaul** [comment]
This requires C++11 and s3fs currently only requires C++03.  We haven't really talked about this before since C++11's benefits are modest but #1068 has discussion about a library we would like to use which requires C++14.

File: src/fdcache_entity.cpp [hunk 1]
```diff
@@ -643,8 +644,8 @@ int FdEntity::Open(const headers_t* pmeta, off_t size, const struct timespec& ts
                 S3FS_PRN_WARN("failed to save cache stat file(%s), but continue...", path.c_str());
             }
         }
-        if(pcfstat){
-            delete pcfstat;
+        if(pcfstat.get() != nullptr){
+            delete pcfstat.release();
```
Comment (gaul) [inline src/fdcache_entity.cpp:648]: This seems to be an unusual way to use `unique_ptr`, instead lifetimes should automatically delete this.

**Review by gaul — CHANGES_REQUESTED** [review]
We are definitely leaking on error paths, probably in other parts of the code as well.  I don't have a strong opposition to this but let's discuss it in the related issue.

**Comment by gaul** [comment]
Alternatively we could use a `unique_ptr` reimplementation that I experimented with in https://github.com/gaul/s3fs-fuse/pull/new/c++/unique_ptr.  However these usually have bugs.

File: src/fdcache_entity.cpp (hunk 1 above)
Comment (eryugey) [reply src/fdcache_entity.cpp]: Removed.

**Comment by eryugey** [comment]
> We are definitely leaking on error paths, probably in other parts of the code as well. I don't have a strong opposition to this but let's discuss it in the related issue.

OK, I switched to auto_ptr this time, and we could switch it back to unique_ptr once s3fs adopts c++11 or above.

**Review by ggtakec — APPROVED** [review]
@eryugey Thanks for the fix code.
_(This code was recently modified and created many leaks.)_

@gaul Please check.
I think we can merge this.

**Comment by gaul** [comment]
Discussing this in #2195.

**Comment by ggtakec** [comment]
@gaul Would you like to close this PR and contain this to #2195?

**Comment by gaul** [comment]
#2195 is a better approach but let's wait until we sort out 1.93.

**Pull request merged**